delfin/drivers/hpe/hpe_3par/ssh_handler.py

```diff
@@ -71,38 +86,358 @@ def get_health_state(self):
+                    'disk_model': str_info[fw_rev_index - 2],
+                    'disk_serial': str_info[fw_rev_index - 1],
+                    'disk_fw_rev': str_info[fw_rev_index]
+                }
+                obj_list.append(inventory_map)
+        return obj_list
+
+    def analyse_nodes(self, cols_size, titles_size, str_info, obj_list):
+        if cols_size >= titles_size:
+            # Only node_ The name attribute may contain spaces,
+            # so there will be several more columns
+            # after splitting
+            # You need to start with the last few columns
+            obj_model = {
+                'node_id': str_info[0],
+                'node_name': ' '.join(str_info[1:cols_size - 8]),
+                'node_state': str_info[cols_size - 8],
+                'node_control_mem': str_info[cols_size - 3],
+                'node_data_mem': str_info[cols_size - 2]
+            }
+            obj_list.append(obj_model)
+        return obj_list
+
+    def analyse_node_cpu(self, cols_size, titles_size, str_info, obj_map):
+        if cols_size >= titles_size:
+            node_id = str_info[0]
+            cpu_info = str_info[4]
+            if obj_map.get(node_id):
+                obj_map[node_id][cpu_info] = obj_map.get(node_id).get(
+                    cpu_info, 0) + 1
+            else:
+                cpu_info_map = {}
+                cpu_info_map[cpu_info] = 1
+                obj_map[node_id] = cpu_info_map
+        return obj_map
+
+    def get_index_of_key(self, titles_list, key):
+        if titles_list:
+            for title in titles_list:
+                if key in title:
+                    return titles_list.index(title)
+        return None
+
+    @staticmethod
+    def do_exec(command_str, ssh):
+        """Execute command"""
+        try:
+            utils.check_ssh_injection(command_str)
+            if command_str is not None and ssh is not None:
+                stdin, stdout, stderr = ssh.exec_command(command_str)
+                res, err = stdout.read(), stderr.read()
+                re = res if res else err
+                result = re.decode()
+        except paramiko.AuthenticationException as ae:
+            LOG.error('doexec Authentication error:{}'.format(ae))
+            raise exception.InvalidUsernameOrPassword()
+        except Exception as e:
+            err = six.text_type(e)
+            LOG.error('doexec InvalidUsernameOrPassword error')
+            if 'timed out' in err:
+                raise exception.SSHConnectTimeout()
+            elif 'No authentication methods available' in err \
+                    or 'Authentication failed' in err:
+                raise exception.InvalidUsernameOrPassword()
+            elif 'not a valid RSA private key file' in err:
+                raise exception.InvalidPrivateKey()
+            else:
+                raise exception.SSHException(err)
+        return result
```

Comment:
This will create exception if 'result' is not defined (if condition on line 397 fails)

Reply:
Modified，In addition, when NetApp is integrated, 3PAR's SSH methods are also modified to use public methods to query